frontend/packages/shared/src/components/TreeView/TreeViewItem/TreeViewItem.tsx

```diff
@@ -0,0 +1,128 @@
+import classes from './TreeViewItem.module.css';
+
+export interface TreeViewItemProps {
+  as?: ElementType;
+  children?: ReactNode;
+  label: ReactNode;
+  labelWrapper?: (children: ReactNode) => ReactNode;
+  nodeId: string;
+}
+
+export const TreeViewItem = ({
+  as = 'li',
+  children,
+  label,
+  labelWrapper = (lab) => lab,
+  nodeId,
+}: TreeViewItemProps) => {
+  const [open, setOpen] = useState(false);
+  const { selectedId, setSelectedId, rootId, focusedId, setFocusedId, focusableId } =
+    useTreeViewRootContext();
+  const { level } = useTreeViewItemContext();
+  const treeItemRef = useRef<HTMLButtonElement>(null);
+
+  useEffect(() => {
+    if (focusedId === nodeId) {
+      treeItemRef.current?.focus();
+    }
+  }, [focusedId, nodeId]);
+
+  const selected = selectedId === nodeId;
+  const focusable = focusableId === nodeId;
+
+  const handleClick = () => {
+    setOpen((prevOpen) => !prevOpen);
+    setSelectedId(nodeId);
+  };
+
+  const handleKeyDown = (event: KeyboardEvent<HTMLButtonElement>) => {
+    switch (event.key) {
+      case 'ArrowRight': // Open node if closed, focus on first child if open, do nothing if not expandable
+        if (children) {
+          open ? setFocusedId(findFirstChildId(rootId, nodeId)) : setOpen(true);
+        }
+        break;
+      case 'ArrowLeft': // Close node if open, focus on parent otherwise
+        open ? setOpen(false) : setFocusedId(findParentId(rootId, nodeId));
+        break;
+      case 'ArrowDown': // Focus on next visible node
+        const nextVisibleNode = findNextVisibleNodeId(rootId, nodeId);
+        if (nextVisibleNode) setFocusedId(nextVisibleNode);
+        break;
+      case 'ArrowUp': // Focus on previous visible node
+        const previousVisibleNode = findPreviousVisibleNodeId(rootId, nodeId);
+        if (previousVisibleNode) setFocusedId(previousVisibleNode);
+        break;
+      case 'Home': // Focus on first node
+        setFocusedId(findFirstNodeId(rootId));
+        break;
+      case 'End': // Focus on last visible node
+        setFocusedId(findLastVisibleNodeId(rootId));
+        break;
+    }
+  };
+
+  const handleFocus = () => setFocusedId(nodeId);
+
+  const treeItemId = makeDomTreeItemId(rootId, nodeId);
+  const listId = makeDomGroupId(rootId, nodeId);
+
+  const renderLabel = () => (
+    <Button
+      aria-expanded={children ? open : undefined}
+      aria-level={level}
+      aria-owns={listId}
+      aria-selected={selected}
+      className={classes.button}
+      id={treeItemId}
+      onClick={handleClick}
+      onFocus={handleFocus}
+      onKeyDown={handleKeyDown}
+      ref={treeItemRef}
+      role='treeitem'
+      size='small'
+      tabIndex={focusable ? 0 : -1}
+      type='button'
+      variant='quiet'
+    >
+      <span aria-hidden className={classes.chevronWrapper}>
+        {children && (open ? <ChevronDownIcon /> : <ChevronRightIcon />)}
+      </span>
+      {label}
+    </Button>
+  );
+
+  const Component = as;
```

Comment:
Bare av ren nysgjerrighet hvorfor brukes ikke li-elementet direkte i komponent-rendringen? Jeg er ikke så dreven på context og provider bruk så antar kanskje det har noe med det å gjøre? 🫣

Reply:
Dette er et konsept som kalles polymorfisme. Det er ganske nytt for meg også, men vi bruker det mye i designsystemet for å øke fleksibiliteten. Kort sagt går det ut på å gjøre det mulig for konsumenten å rendre komponenten som en selvvalgt HTML-tagg. Her er det nødvendig fordi vi vil gjøre det mulig å wrappe komponenten i en annen (f.eks. en dra og slipp-komponent), og da må vi bruke en annen tagg en `li` siden `li` må ligge direkte under `ul`.

Reply:
Okei, stilig konsept!